src/Week7/HW07.hs

```diff
@@ -0,0 +1,197 @@
+{-# LANGUAGE MonadComprehensions, RecordWildCards #-}
+{-# OPTIONS_GHC -Wall #-}
+module Week7.HW07 where
+
+import Prelude hiding (mapM)
+import Week7.Cards
+
+import Control.Monad hiding (mapM, liftM)
+import Control.Monad.Random
+import Data.Functor
+import Data.Monoid
+import Data.Vector (Vector, cons, (!), (!?), (//))
+import System.Random
+
+import qualified Data.Vector as V
+
+
+-- Exercise 1 -----------------------------------------
+
+liftM :: Monad m => (a -> b) -> m a -> m b
+liftM f m =
+    do
+        f <$> m
+
+liftM' :: Monad m => (a -> b) -> m a -> m b
+liftM' f m = f <$> m
+
+swapV :: Int -> Int -> Vector a -> Maybe (Vector a)
+swapV i j vec =
+    liftM2 (\v1 v2 -> vec // [(i,v2), (j,v1)]) (vec !? i) (vec !? j)
+
+-- Exercise 2 -----------------------------------------
+
+mapM :: Monad m => (a -> m b) -> [a] -> m [b]
+mapM f xs = sequence $ f <$> xs
+
+getElts :: [Int] -> Vector a -> Maybe [a]
+getElts xs vec = Week7.HW07.mapM (vec !?) xs
+
+-- Exercise 3 -----------------------------------------
+
+type Rnd a = Rand StdGen a
+
+randomElt :: Vector a -> Rnd (Maybe a)
+randomElt vec =
+    do
+        ind <- getRandomR (0, V.length vec -1)
+        return $ vec !? ind
+
+
+-- Exercise 4 -----------------------------------------
+
+randomVec :: Random a => Int -> Rnd (Vector a)
+randomVec n = V.fromList <$> replicateM n getRandom
+
+randomVecR :: Random a => Int -> (a, a) -> Rnd (Vector a)
+randomVecR n (lo,hi) = V.fromList <$> replicateM n (getRandomR (lo,hi))
+
+-- Exercise 5 -----------------------------------------
+
+swap :: (Int, Int) -> Vector a -> Vector a
+swap (i,j) vec = 
+    let jv = vec ! j in 
+    let iv = vec ! i in 
+    vec // [(i, jv), (j, iv)]
+
+shuffle :: Vector a -> Rnd (Vector a)
+shuffle vec = do
+    swaps <- Week7.HW07.mapM swapPartner [n-1,n-2 .. 1]
+    return $ foldr swap vec swaps
+        where
+            n = V.length vec
+            swapPartner i = (\j -> (i,j)) <$> getRandomR (0,n)
+
+
+-- Exercise 6 -----------------------------------------
+
+partitionAt :: Ord a => Vector a -> Int -> (Vector a, a, Vector a)
+partitionAt vec i = (less, pivotVal, more)
+    where
+        pivotVal = vec ! i
+        (less,more) = V.partition (< pivotVal) vec
+
+
+-- Exercise 7 -----------------------------------------
+
+-- Quicksort
+quicksort :: Ord a => [a] -> [a]
+quicksort [] = []
+quicksort (x:xs) = quicksort [ y | y <- xs, y < x ]
+                   <> (x : quicksort [ y | y <- xs, y >= x ])
+
+qsort :: Ord a => Vector a -> Vector a
+qsort = V.fromList . quicksort . V.toList
+
+
+-- Exercise 8 -----------------------------------------
+
+qsortR :: Ord a => Vector a -> Rnd (Vector a)
+qsortR vec
+    | V.null vec = return V.empty
+    | otherwise = details >>= helperSort
+        where
+            details = getRandomR (0, V.length vec -1) >>= return . partitionAt vec
```

Comment:
instead of 
```haskell
action >>= return . pureFunction
```
it's more common to write
```haskell
fmap pureFunction action
--or 
pureFunction <$> action
```